benchflow-test-manager/application/src/main/java/cloud/benchflow/testmanager/scheduler/running/RunningStatesHandler.java

```diff
@@ -0,0 +1,370 @@
+import cloud.benchflow.testmanager.tasks.running.DetermineExecuteExperimentsTask;
+import cloud.benchflow.testmanager.tasks.running.DetermineExecuteInitialValidationSetTask;
+import cloud.benchflow.testmanager.tasks.running.DetermineExplorationStrategyTask;
+import cloud.benchflow.testmanager.tasks.running.HandleExperimentResultTask;
+import cloud.benchflow.testmanager.tasks.running.RemoveNonReachableExperimentsTask;
+import cloud.benchflow.testmanager.tasks.running.ValidatePredictionFunctionTask;
+import cloud.benchflow.testmanager.tasks.running.ValidateTerminationCriteria;
+import cloud.benchflow.testmanager.tasks.running.ValidateTerminationCriteria.TerminationCriteriaResult;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Jesper Findahl (jesper.findahl@gmail.com) created on 2017-07-02
+ */
+public class RunningStatesHandler {
+
+  /**
+   * Every running state checks if the test has been terminated after each task
+   * to avoid that the test state is overwritten.
+   */
+
+  private static Logger logger =
+      LoggerFactory.getLogger(RunningStatesHandler.class.getSimpleName());
+
+  private ConcurrentMap<String, Future> testTasks;
+
+  private ExecutorService taskExecutorService;
+
+  private ExplorationModelDAO explorationModelDAO;
+  private BenchFlowTestModelDAO testModelDAO;
+
+  private TestTaskScheduler testTaskScheduler;
+
+  public RunningStatesHandler(ConcurrentMap<String, Future> testTasks,
+      ExecutorService taskExecutorService, TestTaskScheduler testTaskScheduler) {
+
+    this.testTasks = testTasks;
+    this.taskExecutorService = taskExecutorService;
+    this.testTaskScheduler = testTaskScheduler;
+  }
+
+  public void initialize() {
+
+    this.explorationModelDAO = BenchFlowTestManagerApplication.getExplorationModelDAO();
+    this.testModelDAO = BenchFlowTestManagerApplication.getTestModelDAO();
+
+  }
+
+  public void determineExplorationStrategy(String testID) {
+
+    logger.info("determineExplorationStrategy with testID: " + testID);
+
+    DetermineExplorationStrategyTask task = new DetermineExplorationStrategyTask(testID);
+
+    // TODO - should go into a stateless queue (so that we can recover)
+    Future<?> future = taskExecutorService.submit(task);
+
+    // replace with new task
+    testTasks.put(testID, future);
+
+    waitForRunningTaskToComplete(testID, future, ADD_STORED_KNOWLEDGE);
+  }
+
+  public void addStoredKnowledge(String testID) {
+
+    logger.info("addStoredKnowledge with testID: " + testID);
+
+    AddStoredKnowledgeTask task = new AddStoredKnowledgeTask(testID);
+
+    // TODO - should go into a stateless queue (so that we can recover)
+    Future<?> future = taskExecutorService.submit(task);
+
+    // replace with new task
+    testTasks.put(testID, future);
+
+    try {
```

Comment:
This can be handled by `waitForRunningTaskToComplete` accepting a callback function to be executed that can contain the following code.